grpcdb/server.go

```diff
@@ -0,0 +1,171 @@
+package grpcdb
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/tendermint/tmlibs/db"
+	protodb "github.com/tendermint/tmlibs/proto"
+)
+
+// ListenAndServe is a blocking function that sets up a gRPC based
+// server at the address supplied, with the gRPC options passed in.
+// Normally in usage, invoke it in a goroutine like you would for http.ListenAndServe.
+func ListenAndServe(addr string, cert string, key string, opts ...grpc.ServerOption) error {
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return err
+	}
+	srv, err := NewServer(cert, key, opts...)
+	if err != nil {
+		return err
+	}
+	return srv.Serve(ln)
+}
+
+func NewServer(cert string, key string, opts ...grpc.ServerOption) (*grpc.Server, error) {
+	creds, err := credentials.NewServerTLSFromFile(cert, key)
+	if err != nil {
+		return nil, err
+	}
+	opts = append(opts, grpc.Creds(creds))
+	srv := grpc.NewServer(opts...)
+	protodb.RegisterDBServer(srv, new(server))
+	return srv, nil
+}
+
+type server struct {
+	mu sync.Mutex
+	db db.DB
+}
+
+var _ protodb.DBServer = (*server)(nil)
+
+// Init initializes the server's database. Only one type of database
+// can be initialized per server.
+//
+// Dir is the directory on the file system in which the DB will be stored(if backed by disk) (TODO: remove)
+//
+// Name is representative filesystem entry's basepath
+//
+// Type can be either one of:
+//  * cleveldb (if built with gcc enabled)
+//  * fsdb
+//  * memdB
+//  * leveldb
+// See https://godoc.org/github.com/tendermint/tmlibs/db#DBBackendType
+func (s *server) Init(ctx context.Context, in *protodb.Init) (*protodb.Entity, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.db = db.NewDB(in.Name, db.DBBackendType(in.Type), in.Dir)
+	return &protodb.Entity{CreatedAt: time.Now().Unix()}, nil
+}
+
+func (s *server) Delete(ctx context.Context, in *protodb.Entity) (*protodb.Nothing, error) {
+	s.db.Delete(in.Key)
+	return nothing, nil
+}
+
+var nothing = new(protodb.Nothing)
+
+func (s *server) DeleteSync(ctx context.Context, in *protodb.Entity) (*protodb.Nothing, error) {
```

Comment:
why return `nothing`?

Reply:
gRPC needs a value, maybe?

Reply:
Yes, the protobuf compiler can't generate an rpc definition that without a return value i.e.
```go
service Foo {
   rpc Bar() 
}
```  
but it'll accept
```go
service Foo {
   rpc Foo(Nothing) returns (Nothing)
}
```